### PR TITLE
LibGfx/ICO: Do not try to decode a mask if we already reached EOF

### LINKED ISSUE #22617: LibGfx: Failed to decode this specific .ico image
This (https://tweakers.net/favicon.ico) favicon image fails to decode, but in Firefox and Chrome it works fine.

This is the log with ICO_DEBUG & BMP_DEBUG enabled:
```
147502.405 ImageDecoder(30294): BMP has invalid magic header number: 0x0000
147502.405 ImageDecoder(30294): load_ico_directory: index 0 width: 64 height: 64 offset: 70 size: 16424
147502.405 ImageDecoder(30294): load_ico_directory: index 1 width: 32 height: 32 offset: 16494 size: 4136
147502.405 ImageDecoder(30294): load_ico_directory: index 2 width: 24 height: 24 offset: 20630 size: 2344
147502.405 ImageDecoder(30294): load_ico_directory: index 3 width: 16 height: 16 offset: 22974 size: 1064
147502.405 ImageDecoder(30294): BMP dib size: 40
147502.405 ImageDecoder(30294): BMP width: 64
147502.405 ImageDecoder(30294): BMP height: 128
147502.405 ImageDecoder(30294): BMP bits_per_pixel: 32
147502.405 ImageDecoder(30294): BMP compression: 0
147502.405 ImageDecoder(30294): BMP image size: 0
147502.405 ImageDecoder(30294): BMP horizontal res: 0
147502.405 ImageDecoder(30294): BMP vertical res: 0
147502.405 ImageDecoder(30294): BMP colors: 0
147502.405 ImageDecoder(30294): BMP important colors: 0
```

### PR DESCRIPTION
**Pull request by LucasChollet**
Fixes #22617

---
**LibGfx/ICO: Do not try to decode a mask if we already reached EOF**

When using the BMP encoding, ICO images are expected to contain a 1-bit
mask for transparency. Regardless an alpha channel is already included
in the image, the mask is always required. As stated here\[1], the
mask is used to provide shadow around the image.

Unfortunately, it seems that some encoder do not include that second
transparency mask. So let's read that mask only if some data is still
remaining after decoding the image.

The test case has been generated by truncating the 64 last bytes
(originally dedicated to the mask) from the `serenity.ico` file and
changing the declared size of the image in the ICO header. The size
value is stored at the offset 0x0E in the file and I changed the value
from 0x0468 to 0x0428.

\[1]: https://devblogs.microsoft.com/oldnewthing/20101021-00/?p=12483

### CODE REVIEW

**Comment by nico** [comment]
> ICO files are always in the ARGB8888

Do you mean after decoding? In the file, I'm pretty sure they aren't always argb8888 (?)

**Comment by LucasChollet** [comment]
You're right, I got confused by one of our comments (and obviously the lack of proper documentation of the format).

File: Userland/Libraries/LibGfx/ImageFormats/BMPLoader.cpp [hunk 1]
```diff
@@ -1429,7 +1429,7 @@ static ErrorOr<void> decode_bmp_pixel_data(BMPLoadingContext& context)
             TRY(process_row(row));
         }
 
-        if (context.is_included_in_ico) {
+        if (context.is_included_in_ico && streamer.remaining() != 0) {
```
Comment (trflynn89) [inline Userland/Libraries/LibGfx/ImageFormats/BMPLoader.cpp:1432]: This is minor, but we have `streamer.at_end()` for this purpose

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/BMPLoader.cpp]: Sure!

**Review by trflynn89 — APPROVED** [review]

**Pull request merged**